apollo-at-link/src/federation_schema.rs

```diff
@@ -0,0 +1,244 @@
+use crate::database::links_metadata;
+use crate::link::LinksMetadata;
+use crate::referencer::{referencers, Referencers};
+use apollo_compiler::Schema;
+use apollo_federation_error::error::FederationError;
+
+pub mod remove;
+
+// A new type that exists purely to implement AsRef and AsMut on them, as we can't do it on
+// Option<LinksMetadata> directly due to the orphan rule.
+pub struct OptionLinksMetadata(Option<LinksMetadata>);
+
+impl From<Option<LinksMetadata>> for OptionLinksMetadata {
+    fn from(value: Option<LinksMetadata>) -> Self {
+        OptionLinksMetadata(value)
+    }
+}
+
+impl AsMut<OptionLinksMetadata> for OptionLinksMetadata {
+    fn as_mut(&mut self) -> &mut OptionLinksMetadata {
+        self
+    }
+}
+
+impl AsRef<OptionLinksMetadata> for OptionLinksMetadata {
+    fn as_ref(&self) -> &OptionLinksMetadata {
+        self
+    }
+}
+
+// Note that LinksMetadata is not automatically updated for all changes to the schema, and it's up
+// to the caller to determine when re-computation is necessary (in the JS code, this metadata was
+// embedded in the schema as "CoreFeatures").
+pub struct FederationSchema {
```

Comment:
shouldn't this just map to existing `Subgraph` object instead?

Reply:
`FederationSchema` here just means a schema with federation metadata attached. It could be a subgraph, but it could also be a supergraph (or even a non-link schema, in which case the metadata is `None`).